events/events.go

```diff
@@ -31,9 +35,83 @@ type EventListener interface {
 	HandleEvent(event Event)
 }
 
+// EventEmitter describes an interface that can be used to emit events
+type EventEmitter interface {
+	EmitEvent(Event)
+}
+
 // EventRegistrar is an object that you can register EventListeners on.
 type EventRegistrar interface {
 	RegisterListener(EventListener)
+	DeregisterListener(EventListener)
+}
+
+type baseEventRegistrar struct {
+	lock      sync.RWMutex
+	listeners []EventListener
+}
+
+// RegisterListener adds a listener to the Event Registar. Events emitted on this
+// registar will be invoked on the listener
+func (a *baseEventRegistrar) RegisterListener(l EventListener) {
+	a.lock.Lock()
+	defer a.lock.Unlock()
+
+	// by making a copy the backing array will never be changed after its creation.
```

Comment:
Is this really necessary or might it be over-optimization? Why is a readlock while iterating not enough?